### PR TITLE
AP_Notify: Neopixel LED driver - NO OREO

### PR DESCRIPTION
**Pull request by magicrub**
Same as https://github.com/ArduPilot/ardupilot/pull/10970 but without the Oreo integration

### CODE REVIEW

**Comment by tridge** [comment]
@magicrub I've reworked this in a few ways and pushed to your branch for you to test. Key changes are:

 - fixed a bug where it mapped to the wrong channel on output
 - allow for up to 32 LEDs in a chain per channel
 - allow RGB to be set much more flexibly
 - only send pulses when we need to, using a separate hal.rcou->neopixel_send() API, allowing a pattern to be first built then sent
 - avoid having to overwrite unchanged bits

Please take a look and let me know how it goes for you. I've tested chains of 8 LEDs with fancy patterns and it works nicely

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 1]
```diff
@@ -1612,4 +1597,67 @@ uint32_t RCOutput::protocol_bitrate(const enum output_mode mode) const
...
+  the given max number of LEDs in the chain.
+*/
+bool RCOutput::set_neopixel_num_LEDs(const uint16_t chan, uint8_t num_leds)
+{
+    uint8_t i;
+    pwm_group *grp = find_chan(chan, i);
+    if (!grp) {
+        return false;
+    }
+    if (grp->neopixel_nleds == num_leds && grp->current_mode == MODE_NEOPIXEL) {
```
Comment (magicrub) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp]: does this need an (... && num_leds>0) check?

Comment (tridge) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp]: it should all work with zero, will end up sending just the pad bits

**Comment by magicrub** [comment]
I won't be able to test this for a while, right in the middle of some non-LED stuff and my LED test bed is in a box somewhere right now. I'll have to trust you that it works, I'll check the rest of the PR for other stuff

**Comment by tridge** [comment]
need to test blheli pass-thru

**Comment by lvale** [comment]
@tridge I might be missing, but how to set the patterns and the conditions for each pattern? Phase 2?

**Comment by tridge** [comment]
@WickedShell I've confirmed blheli passthru is working fine with this change

**Comment by tridge** [comment]
@lvale I'd like to add a Lua scripting interface to control patterns and conditions

**Comment by proficnc** [comment]
Please look at the amazing work that Siddharth has already done for profiLEDs, the pattern generation tool should be a good starting point.

**Comment by auturgy** [comment]
There’s also https://github.com/prickle/Mavpixel
which has a useable GUI

Regards,

James

**Comment by amilcarlucas** [comment]
Is there any documentation at all on how to use this? Google did not find it in the wiki.

**Comment by magicrub** [comment]
@amilcarlucas No, and it's not quite ready. Not sure what to say yet. At this point this implements a primitive driver to display arbitrary LEDs but doesn't know what to display yet. We need to add at least one theme (aka Oreo) and this library was set up to use a new Notify Theme library that is yet to be written. Once a notify theme is implemented then we can document how to use that, and then adding more themes will be more obvious. Idea will be to make it easy to display any arbitrary LED patterns and hopefully make it user-definable, possibly through an app that @bugobliterator wrote a while back for the ProfiLEDs.

**Pull request merged**